### PR TITLE
boards: Add support for Khadas Edge-V board

### PR DESCRIPTION
**Pull request by hx-bora**
This is an AArch64 board. We also add RK3399 SoC support.

See https://www.khadas.com/edge-v for more info about the board.

### CODE REVIEW

File: dts/arm64/rockchip/rockchip_rk3399.dtsi [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_PPI 13 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 14 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 11 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 10 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>;
+		label = "arch_timer";
```
Comment (carlocaione) [inline dts/arm64/rockchip/rockchip_rk3399.dtsi:84]: no need to have the `label` property anywhere anymore.

File: soc/arm64/rockchip/rk3399/soc.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * This needs to be run from the very beginning.
+ * So the init priority has to be 0 (zero).
+ *
+ * @return 0
+ */
+static int rk3399_init(const struct device *arg)
+{
+	ARG_UNUSED(arg);
+
+	/* Nothing yet */
```
Comment (carlocaione) [inline soc/arm64/rockchip/rk3399/soc.c:24]: if nothing is done here can we remove this?

File: boards/arm64/khadas_edgev/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+.. code-block:: console
+
+	mkimage -C none -A arm64 -O linux -a 0x10000000 -e 0x10000000 -d build/zephyr/zephyr.bin build/zephyr/zephyr.img
+
+Use u-boot to load and kick Zephyr.bin to CPU Core0:
+
+.. code-block:: console
+
+	tftpboot ${pxefile_addr_r} zephyr.img; bootm start ${pxefile_addr_r}; bootm loados; bootm go
```
Comment (carlocaione) [inline boards/arm64/khadas_edgev/doc/index.rst:67]: is this correct? No cache flushing?

File: drivers/serial/Kconfig.snps8250dw [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig UART_SNPS8250DW
+	bool "SNPS8250DW serial driver"
+	select SERIAL_HAS_DRIVER
+	help
```
Comment (carlocaione) [inline drivers/serial/Kconfig.snps8250dw:7]: move to `depends on DT_HAS_*_ENABLED`

File: drivers/serial/uart_snps8250dw.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * Copyright (c) 2022 Huawei France Technologies SASU
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/uart.h>
+
+#define DT_DRV_COMPAT snps8250dw
```
Comment (carlocaione) [inline drivers/serial/uart_snps8250dw.c:11]: the compatible should be `snps,8250dw`

File: drivers/serial/uart_snps8250dw.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	.poll_in = uart_snps8250dw_poll_in,
+	.poll_out = uart_snps8250dw_poll_out,
+	.err_check = uart_snps8250dw_err_check,
+};
+
+static int uart_snps8250dw_init(const struct device *dev)
+{
+	(void) dev;
+	return 0;
+}
```
Comment (carlocaione) [inline drivers/serial/uart_snps8250dw.c:65]: no need to have this (and you can use `ARG_UNUSED`)

File: soc/arm64/rockchip/rk3399/mmu_regions.c [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	MMU_REGION_FLAT_ENTRY("GIC",
+			      DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 1),
+			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 1),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE),
+
+	MMU_REGION_ENTRY("UART",
+			      DT_REG_ADDR(DT_INST(0, rockchip_rk3399_uart)),
+			      KADDR_FROM_PADDR_CT(DT_REG_ADDR(DT_INST(0, rockchip_rk3399_uart))),
+			      DT_REG_SIZE(DT_INST(0, rockchip_rk3399_uart)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE),
```
Comment (carlocaione) [inline soc/arm64/rockchip/rk3399/mmu_regions.c:28]: do not map this. Just modify the UART driver to support the device MMIO API (see `include/zephyr/sys/device_mmio.h`).

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.snps_8250dw [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 Huawei France Technologies SASU
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_SNPS_8250DW
```
Comment (galak) [inline drivers/serial/Kconfig.snps_8250dw:4]: does the 16550 driver not work for this?

File: dts/arm64/rockchip/rockchip_rk3399.dtsi [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+			     <GIC_PPI 14 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 11 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 10 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>;
+	};
+
+	uart2: serial@ff1a0000  {
+		compatible = "rockchip,rk3399-uart", "snps,8250dw";
```
Comment (galak) [inline dts/arm64/rockchip/rockchip_rk3399.dtsi:85]: Why isn't this using `snps,dw-apb-uart` as the linux rk3399.dtsi uses?

**Review by galak — DISMISSED** [review]
question about 8250 uart and if we can use 16550 driver, and use the 'snps,dw-apb-uart' compatible that linux rk3399.dtsi uses.

File: dts/arm64/rockchip/rockchip_rk3399.dtsi [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+			device_type = "cpu";
+			compatible = "arm,cortex-a53";
+			reg = <2>;
+		};
+		cpu@3 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a53";
+			reg = <3>;
+		};
+		cpu@4 {
```
Comment (povergoing) [inline dts/arm64/rockchip/rockchip_rk3399.dtsi:38]: Just out of curiosity, generally, a53 and a72 belong to different cluster and thus have different `mpdir` aff formats, for example, a72's `mpdir` are something different like 0x100 0x200 or other formats? But here seems not the case (a72 and a53 are in the same cluster)?

**Comment by galak** [comment]
Also, rename `dts/arm64/rockchip/rockchip_rk3399.dtsi` to `dts/arm64/rockchip/rk3399.dtsi`

**Comment by carlocaione** [comment]
Please, split the serial changes in a different commit.

**Review by carlocaione — DISMISSED** [review]

**Review by carlocaione — APPROVED** [review]

**Comment by carlocaione** [comment]
@povergoing can you take a look?

**Review by povergoing — APPROVED** [review]
LGTM

**Pull request merged**